### PR TITLE
Disallow floating point as partition key

### PR DESCRIPTION
**Pull request by hexiaoting**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Disallow floating point column as partition key
related to : https://github.com/ClickHouse/ClickHouse/issues/18421#event-4147046255

### CODE REVIEW

File: src/Storages/MergeTree/registerStorageMergeTree.cpp [hunk 1]
```diff
@@ -718,6 +718,11 @@ static StoragePtr create(const StorageFactory::Arguments & args)
         ++arg_num;
     }
 
+    DataTypes data_types = metadata.partition_key.data_types;
+    for (size_t i = 0; i < data_types.size(); ++i)
+        if (strncmp(data_types[i]->getFamilyName(), "Float", 5) == 0)
+            throw Exception("Donot support float point as partition key: " + metadata.partition_key.column_names[i], ErrorCodes::BAD_ARGUMENTS);
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/registerStorageMergeTree.cpp:724]: If should be allowed to ATTACH table even with bad partition key
(so the server can startup after version update).

See args.attach.

File: src/Storages/MergeTree/registerStorageMergeTree.cpp [hunk 2]
```diff
@@ -718,6 +718,11 @@ static StoragePtr create(const StorageFactory::Arguments & args)
         ++arg_num;
     }
 
+    DataTypes data_types = metadata.partition_key.data_types;
+    for (size_t i = 0; i < data_types.size(); ++i)
+        if (strncmp(data_types[i]->getFamilyName(), "Float", 5) == 0)
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/registerStorageMergeTree.cpp:723]: strncmp is not the best way to check.
See IDataType.h, isFloat().

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

**Comment by alexey-milovidov** [comment]
@hexiaoting One test should be adjusted.

We have two ways to do it:
1. Introduce a setting `allow_floating_point_partition_key`, 0 by default. Set it to 1 in this test. The setting will be also useful for some special cases.
2. Change `/` operator to `intDiv` in the test.

**Comment by hexiaoting** [comment]
I will add this setting in MergeTreeSetting, not added in system setting.
Because I think this setting is only used at the time of create a Storage. Is that OK?

**Comment by alexey-milovidov** [comment]
Yes, it will be perfect.

**Comment by alexey-milovidov** [comment]
@hexiaoting It is almost ready to merge.

**Comment by hexiaoting** [comment]
> @hexiaoting It is almost ready to merge.

yes, the setting is already added.

**Comment by alexey-milovidov** [comment]
One existing test has to be adjusted...

**Comment by hexiaoting** [comment]
@alexey-milovidov ready to merge

**Pull request merged**